### PR TITLE
Support accessor set input

### LINKED ISSUE #43: Support accessor in `setInput()`
### Background

React components typically use handlers to notify parent about value changes, e.g. ([Checkbox](https://blueprintjs.com/docs/#core/components/checkbox) from Blueprint.js):
```jsx
<Checkbox
  defaultChecked={false}
  onChange={(event) => handle(event.target.checked)}
/>
```

With shiny.react it is possible to use such a component in the following way:
```r
Checkbox(
  defaultChecked = FALSE,
  onChange = JS('(event) => Shiny.setInputValue("checked", event.target.checked)')
)
```

However, it is inconvenient and error-prone to define the `onChange` handler using JS code embedded in a string, so we have
```r
shiny.react::setInput(inputId, argIdx = 1)
```
which can be seen as equivalent to
```js
(...args) => Shiny.setInputValue(inputId, args[argIdx - 1]) // JS uses 0-based indexing
```

### Problem

The `setInput()` function currently only supports sending one of the arguments to Shiny server. It is insufficient when we need to do something with the argument first, like getting the `.target.checked` property in the `Checkbox` example above. It would be great if we could do something like this:
```r
Checkbox(
  defaultChecked = FALSE,
  onChange = setInput("checked", "[0].target.value")
)
```

### Approach

Consider replacing the `argIdx` argument of `setInput` with `accessor` which would behave as in the example above. Suggest other possible API ideas for `setInput` (e.g. have both `argIdx` and `accessor` arguments) and consult with reviewer before implementation.

### Relevant files

1. `R/react.R:174`: definition of `setInput()`
2. `js/src/react/mapReactData.js:88`: JS-side handler

### PR DESCRIPTION
**Pull request by averissimo**
Related issues
- Closes #43.

## Changes description (include a screenshot if applicable!)

- Support accessor as well as an index in `setInput`
- Adds unit tests in R

Peek 2023-01-20 18-55

## Definition of Done checklist
- [x] :spiral_notepad: README, other documentation and code comments that we have is updated with all information related to the change.
- [x] Unit tests added for all new or changed logic.

- [ ] End-to-end tests added for all new or changed logic.

### CODE REVIEW

File: R/react.R [hunk 1]
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
...
+)
+
+#' @describeIn setInput Gets the value from index in argIdx
+#' @export
+#' @examples
+#' setInput("some_id", 2)
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", argIdx = "numeric"),
+  function(inputId, argIdx = 1) {
```
Comment (kamilzyla) [inline R/react.R:215]: Perhaps we shall change to 0-based indexing? It might be confusing that `setInput(id, 1)` and `setInput(id, "[0]")` are equivalent and `setInput()` can be seen as a shorthand for `JS()`, so 0-based indexing might be more natural actually. WDYT?

File: R/react.R [hunk 2]
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
...
-setInput <- function(inputId, argIdx = 1) {
-  ReactData(
-    type = "input", id = inputId, argIdx = argIdx - 1
-  )
-}
+#' @examples
+#' setInput("some_id")
+methods::setGeneric(
+  "setInput",
+  function(inputId, argIdx = 1) {
```
Comment (kamilzyla) [inline R/react.R:190]: I'm no expert on S4, but this default value seems not to be used?
```suggestion
  function(inputId, argIdx) {
```

File: tests/testthat/test-setInput.R [hunk 3]
```diff
@@ -0,0 +1,37 @@
+library(testthat)
```
Comment (kamilzyla) [inline tests/testthat/test-setInput.R:1]: Kudos for adding the firsts R tests! :grin: Let's remove `test-all.R`. We can also get rid of `mockery` from `DESCRIPTION`.

File: js/webpack.config.babel.js [hunk 4]
```diff
@@ -5,6 +5,7 @@ export default {
   output: {
     path: join(__dirname, '..', 'inst', 'www', 'shiny.react'),
     filename: 'shiny-react.js',
+    hashFunction: "sha256",
```
Comment (kamilzyla) [inline js/webpack.config.babel.js:8]: Can you tell me more, why is this necessary? Ideally we'd fix your setup so it works without this line and perhaps update docs. If that's unpractical or we figure we should actually explicitly set `hashFunction`, please use single quotes (consistency) and add a comment (why such setting).

File: js/src/react/mapReactData.js [hunk 5]
```diff
@@ -85,9 +85,10 @@ dataMappers.element = ({ module, name, props: propsData }) => {
 // Used to implement `setInput()` and `triggerEvent()` R functions. In case of `triggerEvent()`,
 // we have `argIdx === null` and the returned function just sets the Shiny input to `TRUE`
 // on every call (this works thanks to `priority: 'event'`).
-dataMappers.input = ({ id, argIdx }) => (
+dataMappers.input = ({ id, argIdx, accessor }) => (
   (...args) => {
-    const value = argIdx === null ? true : args[argIdx];
+    let value = argIdx === null || argIdx === undefined ? true : args[argIdx];
+    value = accessor === null || accessor === undefined ? value : eval(`${args}accessor`); // eslint-disable-line no-eval
```
Comment (kamilzyla) [inline js/src/react/mapReactData.js:91]: I find this hard to read. Let's have a separate data mapper for `triggerEvent()` to avoid `=== null` checks. Then consider a single `accessor` argument and check its type, or perhaps have separate data mappers for `argIdx` and `accessor` cases.

**Review by kamilzyla — CHANGES_REQUESTED** [review]
Very neat PR!

File: R/react.R [hunk 6]
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
...
+}
+
 #' Set input
 #'
 #' Creates a handler which can be used for `onChange` and similar props of 'React' components
 #' to set the value of a 'Shiny' input to one of the arguments passed to the handler.
 #'
 #' @param inputId 'Shiny' input ID to set the value on.
-#' @param argIdx Index of the argument to use as value.
+#' @param argIdx Index (numeric) or accessor (string) of the argument to use as value.
```
[the remark this answers is not in the document]
Comment (averissimo) [reply R/react.R:182]: I kept the same argument name due to backwards compatibility for the cases where it was called explicitly `setInput('some_id', argIdx = 1)`

If you think we can move past the name, then I agree with a name change to accessor. 

I'll add the extra examples

File: R/react.R [hunk 7]
```diff
@@ -161,18 +161,82 @@ triggerEvent <- function(inputId) {
   )
 }
 
+#' @title usage for function setInput
+#'
+#' @description  Prints a description of valid parameters for the `setInput`
+#' function
+setInputUsage <- function() {
```
[the remark this answers is not in the document]
Comment (averissimo) [reply R/react.R:168]: Will do!

File: js/webpack.config.babel.js (hunk 4 above)
Comment (averissimo) [reply js/webpack.config.babel.js:8]: From my understanding openssl removed support for the hashing algorithm _(md4)_ that older versions of webpack default to. In particular, with the one that's currently used by `shiny.react`.

So when running webpack it just fails.

image

(this is what happens when I run `yarn && yarn webpack` in master)

See [here](https://stackoverflow.com/a/69476335/302485)

If we update webpack to v5.54.0+ this problem goes away and we can remove this output option

**I just didn't want to update node packages along with this PR, but if you're ok with it I can.**

File: js/src/react/mapReactData.js (hunk 5 above)
Comment (averissimo) [reply js/src/react/mapReactData.js:91]: I preferred using a more explicit if conditional, but that was giving me a weird issue, where `args` became `undefined`.

I'll improve on this

File: js/webpack.config.babel.js (hunk 4 above)
Comment (averissimo) [reply js/webpack.config.babel.js]: I've updated the webpack and this solves the problem

File: js/src/react/mapReactData.js (hunk 5 above)
Comment (averissimo) [reply js/src/react/mapReactData.js]: I've unified the solution, so that `jsAccessor = 1` resolves to "jsAccessor[1]" in R. This keeps the code cleaner with a single point of maintenance.

`triggerEvent` has its own dataMapper that is straightforward

**Comment by averissimo** [comment]
I believe I resolved all your comments @kamilzyla 

- removed setInputUsage
- updated webpack so that there's no hashing option to maintain
- rename argIdx to jsAccessor
- break triggerEvent to its own dataMapper called `dataMappers.event`

Please take a look at the changes I did to `setInput` as it seemed redundant to have code in R and JS handling the index vs. accessor

I'm a bit ashamed of the time it took to figure out a bug I was having with `(...args)` to `arguments`

File: R/react.R [hunk 8]
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
...
+)
+
+#' @describeIn setInput Gets the value from index in jsAccessor
+#' @export
+#' @examples
+#' setInput("some_id", 2)
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", jsAccessor = "numeric"),
+  function(inputId, jsAccessor = 1) {
```
Comment (kamilzyla) [inline R/react.R:215]: Is this default value used in any way? If it is, make it a 0 :wink:
```suggestion
  function(inputId, jsAccessor) {
```

File: R/react.R [hunk 9]
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
...
+#' @examples
+#' setInput("some_id", ".target.value")
+#'
+#' setInput("some_id", "[0]")
+#' # which is equivalent to
+#' setInput("some_id", 0)
+methods::setMethod(
+  "setInput",
+  signature(inputId = "character", jsAccessor = "character"),
+  function(inputId, jsAccessor = NULL) {
```
Comment (kamilzyla) [inline R/react.R:240]: Is this default value used in any way?
```suggestion
  function(inputId, jsAccessor) {
```

File: R/react.R [hunk 10]
```diff
@@ -157,22 +157,89 @@ JS <- function(...) { # nolint
...
+    type = "event", id = inputId
   )
 }
 
-#' Set input
+#' @title Set input
 #'
-#' Creates a handler which can be used for `onChange` and similar props of 'React' components
-#' to set the value of a 'Shiny' input to one of the arguments passed to the handler.
+#' @description Creates a handler which can be used for `onChange` and similar
```
Comment (kamilzyla) [inline R/react.R:166]: I'd suggest to drop `@title` and `@description`:
1. Consistency: that's how the rest of our docs are written.
2. The first paragraph will become the `@description` and the remaining ones (starting with "The argument...") will become `@details` which makes sense in the rendered docs.

File: js/src/react/mapReactData.js [hunk 11]
```diff
@@ -82,12 +82,25 @@ dataMappers.element = ({ module, name, props: propsData }) => {
...
+    //
+    let value = true;
+    if (jsAccessor !== undefined) {
+      // Needs to use arguments inside eval string, otherwise webpack
+      // won't translate args => arguments
+      value = eval(`arguments${jsAccessor}`); // eslint-disable-line no-eval
+    }
     Shiny.setInputValue(id, value, { priority: 'event' });
   }
 );
```
Comment (kamilzyla) [inline js/src/react/mapReactData.js:106]: WDYT about this implementation? I dropped the `jsAccessor !== undefined` check - when could it fail?
```js
dataMappers.input = ({ id, jsAccessor }) => {
  const getValue = eval(`(...args) => (args${jsAccessor})`); // eslint-disable-line no-eval
  return (...args) => {
    Shiny.setInputValue(id, getValue(args), { priority: 'event' });
  }
}
```

**Review by kamilzyla — CHANGES_REQUESTED** [review]
Great job! Have a bunch more comments, mostly nitpicks about documentation :wink:

File: js/src/react/mapReactData.js (hunk 11 above)
Comment (kamilzyla) [reply js/src/react/mapReactData.js:106]: Note: There was a bug in my suggested code above (`...args` instead of `args` in the `eval()`). Now fixed.

Comment (averissimo) [reply js/src/react/mapReactData.js]: Very clean!!

I like it a lot and it avoids having to use `arguments` directly as it's no longer inside of a string

I like to safeguard against some stupid bugs, but here it's really not necessary

File: R/react.R (hunk 10 above)
Comment (averissimo) [reply R/react.R]: Done!

**Comment by averissimo** [comment]
I dropped the ball on the documentation. Thank you for catching it!

**Review by kamilzyla — APPROVED** [review]
LGTM :+1: Great job and thanks for your patience :smiley: :tada: Merging with a tiny docs change of mine.
